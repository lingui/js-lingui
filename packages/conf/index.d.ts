import type { GeneratorOptions } from "@babel/core"

export declare type CatalogFormat =
  | "lingui"
  | "minimal"
  | "po"
  | "csv"
  | "po-gettext"
export type CatalogFormatOptions = {
  origins?: boolean
  lineNumbers?: boolean
}
export declare type OrderBy = "messageId" | "origin"
declare type CatalogConfig = {
  name?: string
  path: string
  include: string[]
  exclude?: string[]
}

export type LocaleObject = {
  [locale: string]: string[] | string
}

export type DefaultLocaleObject = {
  default: string
}

export declare type FallbackLocales = LocaleObject | DefaultLocaleObject

declare type CatalogService = {
  name: string
  apiKey: string
}

declare type ExtractorType = {
  match(filename: string): boolean
  extract(filename: string, targetDir: string, options?: any): void
}

export declare type LinguiConfig = {
  catalogs: CatalogConfig[]
  compileNamespace: "es" | "cjs" | "ts" | string
  extractBabelOptions: Record<string, unknown>
  compilerBabelOptions: GeneratorOptions
  fallbackLocales: FallbackLocales
  format: CatalogFormat
  extractors?: ExtractorType[]
  prevFormat: CatalogFormat
  formatOptions: CatalogFormatOptions
  localeDir: string
  locales: string[]
  catalogsMergePath: string
  orderBy: OrderBy
  pseudoLocale: string
  rootDir: string
  runtimeConfigModule:
    | [source: string, identifier?: string]
    | {
        i18n?: [source: string, identifier?: string]
        Trans?: [source: string, identifier?: string]
      }
  sourceLocale: string
  service: CatalogService
}
export declare const defaultConfig: LinguiConfig
export declare function getConfig({
  cwd,
  configPath,
  skipValidation,
}?: {
  cwd?: string
  configPath?: string
  skipValidation?: boolean
}): LinguiConfig

export declare function makeConfig(
  userConfig: Partial<LinguiConfig>,
  opts?: {
    skipValidation?: boolean
  }
): LinguiConfig

export declare const configValidation: {
  exampleConfig: {
    extractBabelOptions: {
      extends: string
      rootMode: string
      plugins: string[]
      presets: string[]
    }
    compilerBabelOptions: GeneratorOptions
    catalogs: CatalogConfig[]
    compileNamespace: "es" | "ts" | "cjs" | string
    fallbackLocales: FallbackLocales
    format: CatalogFormat
    formatOptions: CatalogFormatOptions
    locales: string[]
    orderBy: OrderBy
    pseudoLocale: string
    rootDir: string
    runtimeConfigModule: LinguiConfig["runtimeConfigModule"]
    sourceLocale: string
    service: CatalogService
  }
  deprecatedConfig: {
    fallbackLocale: (
      config: LinguiConfig & DeprecatedFallbackLanguage
    ) => string
    localeDir: (config: LinguiConfig & DeprecatedLocaleDir) => string
    srcPathDirs: (config: LinguiConfig & DeprecatedLocaleDir) => string
    srcPathIgnorePatterns: (
      config: LinguiConfig & DeprecatedLocaleDir
    ) => string
  }
  comment: string
}
export declare function replaceRootDir(
  config: LinguiConfig,
  rootDir: string
): LinguiConfig
/**
 * Replace fallbackLocale with fallbackLocales
 *
 * Released in lingui-conf 0.9
 * Remove anytime after 4.x
 */
declare type DeprecatedFallbackLanguage = {
  fallbackLocale: string | null
}

export declare function fallbackLanguageMigration(
  config: LinguiConfig & DeprecatedFallbackLanguage
): LinguiConfig
/**
 * Replace localeDir, srcPathDirs and srcPathIgnorePatterns with catalogs
 *
 * Released in @lingui/conf 3.0
 * Remove anytime after 4.x
 */
declare type DeprecatedLocaleDir = {
  localeDir: string
  srcPathDirs: string[]
  srcPathIgnorePatterns: string[]
}
export declare function catalogMigration(
  config: LinguiConfig & DeprecatedLocaleDir
): LinguiConfig
export {}
//# sourceMappingURL=index.d.ts.map
