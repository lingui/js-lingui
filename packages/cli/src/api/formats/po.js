// @flow
import fs from "fs"
import * as R from "ramda"
import { format as formatDate } from "date-fns"

import PO from "pofile"
import { joinOrigin, splitOrigin } from "../utils"
import type { TranslationsFormat } from "../types"

const getCreateHeaders = (language = "no") => ({
  "POT-Creation-Date": formatDate(new Date(), "YYYY-MM-DD HH:mmZZ"),
  "Mime-Version": "1.0",
  "Content-Type": "text/plain; charset=utf-8",
  "Content-Transfer-Encoding": "8bit",
  "X-Generator": "@lingui/cli",
  Language: language
})

const serialize = R.compose(
  R.values,
  R.mapObjIndexed((message, key) => {
    const item = new PO.Item()
    item.msgid = key
    item.msgstr = message.translation
    item.comments = message.comments || []
    item.extractedComments = message.description ? [message.description] : []
    item.references = message.origin ? message.origin.map(joinOrigin) : []
    item.obsolete = message.obsolete
    item.flags = message.flags
      ? R.fromPairs(message.flags.map(flag => [flag, true]))
      : {}
    return item
  })
)

const getMessageKey = R.prop("msgid")
const getTranslations = R.prop("msgstr")
const getExtractedComments = R.prop("extractedComments")
const getTranslatorComments = R.prop("comments")
const getOrigins = R.prop("references")
const getFlags = R.compose(
  R.map(R.trim),
  R.keys,
  R.dissoc("obsolete"), // backward-compatibility, remove in 3.x
  R.prop("flags")
)
const isObsolete = R.either(R.path(["flags", "obsolete"]), R.prop("obsolete"))

const deserialize = R.map(
  R.applySpec({
    translation: R.compose(
      R.head,
      R.defaultTo([]),
      getTranslations
    ),
    description: R.compose(
      R.head,
      R.defaultTo([]),
      getExtractedComments
    ),
    comments: item =>
      R.concat(getTranslatorComments(item), R.tail(getExtractedComments(item))),
    obsolete: isObsolete,
    origin: R.compose(
      R.map(splitOrigin),
      R.defaultTo([]),
      getOrigins
    ),
    flags: getFlags
  })
)

const validateItems = R.map(item => {
  if (R.length(getTranslations(item)) > 1) {
    console.warn(
      "Multiple translations for item with key %s is not supported and it will be ignored.",
      getMessageKey(item)
    )
  }
})

const indexItems = R.indexBy(getMessageKey)

const format: TranslationsFormat = {
  filename: "messages.po",

  write(filename, catalog, options = {}) {
    let po
    let indexedItems = {}
    if (fs.existsSync(filename)) {
      const raw = fs.readFileSync(filename).toString()
      po = PO.parse(raw)
    } else {
      po = new PO()
      po.headers = getCreateHeaders(options.language)
      po.headerOrder = R.keys(po.headers)
    }
    po.items = serialize(catalog)
    fs.writeFileSync(filename, po.toString())
  },

  read(filename) {
    const raw = fs.readFileSync(filename).toString()
    return this.parse(raw)
  },

  parse(raw) {
    const po = PO.parse(raw)
    validateItems(po.items)
    return deserialize(indexItems(po.items))
  }
}

export default format
