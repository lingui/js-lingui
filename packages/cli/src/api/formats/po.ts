import fs from "fs"
import * as R from "ramda"
import { format as formatDate } from "date-fns"
import PO from "pofile"

import { joinOrigin, splitOrigin, writeFileIfChanged } from "../utils"
import { MessageType, CatalogType } from "../types"
import { CatalogFormatter } from "./types"

const getCreateHeaders = (language = "no") => ({
  "POT-Creation-Date": formatDate(new Date(), "yyyy-MM-dd HH:mmxxxx"),
  "Mime-Version": "1.0",
  "Content-Type": "text/plain; charset=utf-8",
  "Content-Transfer-Encoding": "8bit",
  "X-Generator": "@lingui/cli",
  Language: language,
})

const serialize = (items: CatalogType, options) =>
  R.compose(
    R.values,
    R.mapObjIndexed((message: MessageType, key) => {
      const item = new PO.Item()
      item.msgid = key
      item.msgstr = [message.translation]
      item.comments = message.comments || []
      item.extractedComments = message.comment ? [message.comment] : []
      if (options.origins) {
        item.references = message.origin ? message.origin.map(joinOrigin) : []
      }
      // @ts-ignore: Figure out how to set this flag
      item.obsolete = message.obsolete
      item.flags = message.flags
        ? R.fromPairs(message.flags.map((flag) => [flag, true]))
        : {}
      return item
    })
  )(items)

const getMessageKey = R.prop<"msgid", string>("msgid")
const getTranslations = R.prop("msgstr")
const getExtractedComments = R.prop("extractedComments")
const getTranslatorComments = R.prop("comments")
const getOrigins = R.prop("references")
const getFlags = R.compose(
  R.map(R.trim),
  R.keys,
  R.dissoc("obsolete"), // backward-compatibility, remove in 3.x
  R.prop("flags")
)
const isObsolete = R.either(R.path(["flags", "obsolete"]), R.prop("obsolete"))

const deserialize: (item: Object) => Object = R.map(
  R.applySpec({
    translation: R.compose(R.head, R.defaultTo([]), getTranslations),
    comment: R.compose(R.head, R.defaultTo([]), getExtractedComments),
    comments: (item) =>
      R.concat(
        getTranslatorComments(item) as string,
        R.tail(getExtractedComments(item))
      ),
    obsolete: isObsolete,
    origin: R.compose(R.map(splitOrigin), R.defaultTo([]), getOrigins),
    flags: getFlags,
  })
)

type POItem = InstanceType<typeof PO.Item>

const validateItems = R.forEach<POItem>((item) => {
  if (R.length(getTranslations(item)) > 1) {
    console.warn(
      "Multiple translations for item with key %s is not supported and it will be ignored.",
      getMessageKey(item)
    )
  }
})

const indexItems = R.indexBy(getMessageKey)

const po: CatalogFormatter = {
  catalogExtension: ".po",

  write(filename, catalog, options) {
    let po
    if (fs.existsSync(filename)) {
      const raw = fs.readFileSync(filename).toString()
      po = PO.parse(raw)
    } else {
      po = new PO()
      po.headers = getCreateHeaders(options.locale)
      po.headerOrder = R.keys(po.headers)
    }
    po.items = serialize(catalog, options)
    writeFileIfChanged(filename, po.toString())
  },

  read(filename) {
    const raw = fs.readFileSync(filename).toString()
    return this.parse(raw)
  },

  // @ts-ignore
  parse(raw) {
    const po = PO.parse(raw)
    validateItems(po.items)
    return deserialize(indexItems(po.items))
  },
}

export default po
