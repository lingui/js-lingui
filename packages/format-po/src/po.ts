import { format as formatDate } from "date-fns"
import PO from "pofile"

import { CatalogFormatter, CatalogType, MessageType } from "@lingui/conf"
import { generateMessageId } from "@lingui/message-utils/generateMessageId"

type POItem = InstanceType<typeof PO.Item>

const splitOrigin = (origin: string) => {
  const [file, line] = origin.split(":")
  return [file, line ? Number(line) : null] as [file: string, line: number]
}

/**
 * @internal
 */
export type POCatalogExtra = {
  translatorComments?: string[]
  flags?: string[]
}

const joinOrigin = (origin: [file: string, line?: number]): string =>
  origin.join(":")

export type PoFormatterOptions = {
  /**
   * Print places where message is used
   *
   * @default true
   */
  origins?: boolean

  /**
   * Print line numbers in origins
   *
   * @default true
   */
  lineNumbers?: boolean

  /**
   * Print `js-lingui-id: Xs4as` statement in extracted comments section
   *
   * @default false
   */
  printLinguiId?: boolean

  /**
   * By default, the po-formatter treats the pair `msgid` + `msgctx` as the source
   * for generating an ID by hashing its value.
   *
   * For messages with explicit IDs, the formatter adds a special comment `js-lingui-explicit-id` as a flag.
   * When this flag is present, the formatter will use the `msgid` as-is without any additional processing.
   *
   * Set this option to true if you exclusively use explicit-ids in your project.
   *
   * https://lingui.dev/tutorials/react-patterns#using-custom-id
   *
   * @default false
   */
  explicitIdAsDefault?: boolean
}

function isGeneratedId(id: string, message: MessageType): boolean {
  return id === generateMessageId(message.message, message.context)
}

function getCreateHeaders(language: string): PO["headers"] {
  return {
    "POT-Creation-Date": formatDate(new Date(), "yyyy-MM-dd HH:mmxxxx"),
    "MIME-Version": "1.0",
    "Content-Type": "text/plain; charset=utf-8",
    "Content-Transfer-Encoding": "8bit",
    "X-Generator": "@lingui/cli",
    ...(language ? { Language: language } : {}),
  }
}

const EXPLICIT_ID_FLAG = "js-lingui-explicit-id"
const GENERATED_ID_FLAG = "js-lingui-generated-id"

const serialize = (catalog: CatalogType, options: PoFormatterOptions) => {
  return Object.keys(catalog).map((id) => {
    const message: MessageType<POCatalogExtra> = catalog[id]

    const item = new PO.Item()

    // The extractedComments array may be modified in this method,
    // so create a new array with the message's elements.
    item.extractedComments = [...(message.comments || [])]

    item.flags = ((message.extra?.flags || []) as string[]).reduce<
      Record<string, boolean>
    >((acc, flag) => {
      acc[flag] = true
      return acc
    }, {})

    const _isGeneratedId = isGeneratedId(id, message)

    if (_isGeneratedId) {
      item.msgid = message.message

      if (options.explicitIdAsDefault) {
        if (!item.extractedComments.includes(GENERATED_ID_FLAG)) {
          item.extractedComments.push(GENERATED_ID_FLAG)
        }
      }

      if (options.printLinguiId) {
        if (!item.extractedComments.find((c) => c.includes("js-lingui-id"))) {
          item.extractedComments.push(`js-lingui-id: ${id}`)
        }
      }
    } else {
      if (!options.explicitIdAsDefault) {
        if (!item.extractedComments.includes(EXPLICIT_ID_FLAG)) {
          item.extractedComments.push(EXPLICIT_ID_FLAG)
        }
      }

      item.msgid = id
    }

    if (message.context) {
      item.msgctxt = message.context
    }

    item.msgstr = [message.translation]
    item.comments = message.extra?.translatorComments || []

    if (options.origins !== false) {
      if (message.origin && options.lineNumbers === false) {
        item.references = message.origin.map(([path]) => path)
      } else {
        item.references = message.origin ? message.origin.map(joinOrigin) : []
      }
    }
    item.obsolete = message.obsolete

    return item
  })
}

function deserialize(
  items: POItem[],
  options: PoFormatterOptions
): CatalogType {
  return items.reduce<CatalogType<POCatalogExtra>>((catalog, item) => {
    const message: MessageType<POCatalogExtra> = {
      translation: item.msgstr[0],
      comments: item.extractedComments || [],
      context: item.msgctxt ?? null,
      obsolete: item.flags.obsolete || item.obsolete,
      origin: (item.references || []).map((ref) => splitOrigin(ref)),
      extra: {
        translatorComments: item.comments || [],
        flags: Object.keys(item.flags).map((flag) => flag.trim()),
      },
    }

    let id = item.msgid

    // if generated id, recreate it
    if (
      options.explicitIdAsDefault
        ? item.extractedComments.includes(GENERATED_ID_FLAG)
        : !item.extractedComments.includes(EXPLICIT_ID_FLAG)
    ) {
      id = generateMessageId(item.msgid, item.msgctxt)
      message.message = item.msgid
    }

    catalog[id] = message
    return catalog
  }, {})
}

export function formatter(options: PoFormatterOptions = {}): CatalogFormatter {
  options = {
    origins: true,
    lineNumbers: true,
    ...options,
  }

  return {
    catalogExtension: ".po",
    templateExtension: ".pot",

    parse(content): CatalogType {
      const po = PO.parse(content)
      return deserialize(po.items, options)
    },

    serialize(catalog, ctx): string {
      let po: PO

      if (ctx.existing) {
        po = PO.parse(ctx.existing)
      } else {
        po = new PO()
        po.headers = getCreateHeaders(ctx.locale)
        // accessing private property
        ;(po as any).headerOrder = Object.keys(po.headers)
      }

      po.items = serialize(catalog, options)
      return po.toString()
    },
  }
}
