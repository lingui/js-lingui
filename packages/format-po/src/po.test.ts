import fs from "fs"
import path from "path"

import { formatter as createFormatter, POCatalogExtra } from "./po"
import { CatalogFormatter, CatalogType } from "@lingui/conf"
import MockDate from "mockdate"

const defaultParseCtx: Parameters<CatalogFormatter["parse"]>[1] = {
  locale: "en",
  sourceLocale: "en",
  filename: "file.po",
}

const defaultSerializeCtx: Parameters<CatalogFormatter["serialize"]>[1] = {
  locale: "en",
  existing: null,
  filename: "file.po",
  sourceLocale: "en",
}

describe("pofile format", () => {
  beforeAll(() => {
    MockDate.set(new Date("2018-08-27T10:00Z"))
  })

  afterAll(() => {
    MockDate.reset()
  })
  it("should write catalog in pofile format", () => {
    const format = createFormatter({ origins: true })

    const catalog: CatalogType<POCatalogExtra> = {
      static: {
        translation: "Static message",
      },
      withOrigin: {
        translation: "Message with origin",
        origin: [["src/App.js", 4]],
      },
      withContext: {
        translation: "Message with context",
        context: "my context",
      },
      Dgzql1: {
        message: "with generated id",
        translation: "",
        context: "my context",
      },
      withMultipleOrigins: {
        translation: "Message with multiple origin",
        origin: [
          ["src/App.js", 4],
          ["src/Component.js", 2],
        ],
      },
      withDescription: {
        translation: "Message with description",
        comments: ["Description is comment from developers to translators"],
      },
      withComments: {
        extra: {
          translatorComments: [
            "Translator comment",
            "This one might come from developer",
          ],
        },
        translation: "Support translator comments separately",
      },
      obsolete: {
        translation: "Obsolete message",
        obsolete: true,
      },
      withFlags: {
        extra: {
          flags: ["fuzzy", "otherFlag"],
        },
        translation: "Keeps any flags that are defined",
      },
      veryLongString: {
        translation:
          "One morning, when Gregor Samsa woke from troubled dreams, he found himself" +
          " transformed in his bed into a horrible vermin. He lay on his armour-like" +
          " back, and if he lifted his head a little he could see his brown belly," +
          " slightly domed and divided by arches into stiff sections. The bedding was" +
          " hardly able to cover it and seemed ready to slide off any moment. His many" +
          " legs, pitifully thin compared with the size of the rest of him, waved about" +
          " helplessly as he looked. \"What's happened to me?\" he thought. It wasn't" +
          " a dream. His room, a proper human",
      },
    }

    const pofile = format.serialize(catalog, defaultSerializeCtx)
    expect(pofile).toMatchSnapshot()
  })

  it("should read catalog in pofile format", () => {
    const format = createFormatter()

    const pofile = fs
      .readFileSync(path.join(__dirname, "fixtures/messages.po"))
      .toString()

    const actual = format.parse(pofile, defaultParseCtx)
    expect(actual).toMatchSnapshot()
  })

  it("should serialize and deserialize messages with generated id", () => {
    const format = createFormatter({ origins: true })

    const catalog: CatalogType = {
      // with generated id
      Dgzql1: {
        message: "with generated id",
        translation: "",
        context: "my context",
      },
    }

    const serialized = format.serialize(catalog, defaultSerializeCtx) as string

    const actual = format.parse(serialized, defaultParseCtx)
    expect(actual).toMatchObject(catalog)
  })

  describe("explicitIdAsDefault", () => {
    const catalog: CatalogType = {
      // with generated id
      Dgzql1: {
        message: "with generated id",
        translation: "",
        context: "my context",
      },

      "custom.id": {
        message: "with explicit id",
        translation: "",
      },
    }

    it("should set `js-lingui-generated-id` for messages with generated id when [explicitIdAsDefault: true]", () => {
      const format = createFormatter({
        origins: true,
        explicitIdAsDefault: true,
      })

      const serialized = format.serialize(
        catalog,
        defaultSerializeCtx
      ) as string

      expect(serialized).toMatchInlineSnapshot(`
        msgid ""
        msgstr ""
        "POT-Creation-Date: 2018-08-27 10:00+0000\\n"
        "MIME-Version: 1.0\\n"
        "Content-Type: text/plain; charset=utf-8\\n"
        "Content-Transfer-Encoding: 8bit\\n"
        "X-Generator: @lingui/cli\\n"
        "Language: en\\n"

        #. js-lingui-generated-id
        msgctxt "my context"
        msgid "with generated id"
        msgstr ""

        msgid "custom.id"
        msgstr ""

      `)

      const actual = format.parse(serialized, defaultParseCtx)
      expect(actual).toMatchInlineSnapshot(`
        {
          Dgzql1: {
            comments: [
              js-lingui-generated-id,
            ],
            context: my context,
            extra: {
              flags: [],
              translatorComments: [],
            },
            message: with generated id,
            obsolete: false,
            origin: [],
            translation: ,
          },
          custom.id: {
            comments: [],
            context: null,
            extra: {
              flags: [],
              translatorComments: [],
            },
            obsolete: false,
            origin: [],
            translation: ,
          },
        }
      `)
    })

    it("should set `js-explicit-id` for messages with explicit id when [explicitIdAsDefault: false]", () => {
      const format = createFormatter({
        origins: true,
        explicitIdAsDefault: false,
      })

      const serialized = format.serialize(
        catalog,
        defaultSerializeCtx
      ) as string

      expect(serialized).toMatchInlineSnapshot(`
        msgid ""
        msgstr ""
        "POT-Creation-Date: 2018-08-27 10:00+0000\\n"
        "MIME-Version: 1.0\\n"
        "Content-Type: text/plain; charset=utf-8\\n"
        "Content-Transfer-Encoding: 8bit\\n"
        "X-Generator: @lingui/cli\\n"
        "Language: en\\n"

        msgctxt "my context"
        msgid "with generated id"
        msgstr ""

        #. js-lingui-explicit-id
        msgid "custom.id"
        msgstr ""

      `)

      const actual = format.parse(serialized, defaultParseCtx)
      expect(actual).toMatchInlineSnapshot(`
        {
          Dgzql1: {
            comments: [],
            context: my context,
            extra: {
              flags: [],
              translatorComments: [],
            },
            message: with generated id,
            obsolete: false,
            origin: [],
            translation: ,
          },
          custom.id: {
            comments: [
              js-lingui-explicit-id,
            ],
            context: null,
            extra: {
              flags: [],
              translatorComments: [],
            },
            obsolete: false,
            origin: [],
            translation: ,
          },
        }
      `)
    })
  })

  it("should print lingui id if printLinguiId = true", () => {
    const format = createFormatter({ origins: true, printLinguiId: true })

    const catalog: CatalogType = {
      // with generated id
      Dgzql1: {
        message: "with generated id",
        translation: "",
        context: "my context",
      },
    }

    const serialized = format.serialize(catalog, defaultSerializeCtx)

    expect(serialized).toMatchSnapshot()
  })

  it("should not add lingui id more than one time", () => {
    const format = createFormatter({ origins: true, printLinguiId: true })

    const catalog: CatalogType = {
      // with generated id
      Dgzql1: {
        message: "with generated id",
        translation: "",
        context: "my context",
        comments: ["js-lingui-id: Dgzql1"],
      },
    }

    const serialized = format.serialize(catalog, defaultSerializeCtx)

    expect(serialized).toMatchSnapshot()
  })

  it("should correct badly used comments", () => {
    const format = createFormatter()

    const po = `
      #. First description
      #. Second comment
      #. Third comment
      #. js-lingui-explicit-id
      msgid "withMultipleDescriptions"
      msgstr "Extra comments are separated from the first description line"

      # Translator comment
      #. Single description only
      #. Second description?
      #. js-lingui-explicit-id
      msgid "withDescriptionAndComments"
      msgstr "Second description joins translator comments"
    `

    const actual = format.parse(po, defaultParseCtx)
    expect(actual).toMatchSnapshot()
  })

  it("should not include origins if origins option is false", () => {
    const format = createFormatter({ origins: false })

    const catalog: CatalogType = {
      static: {
        translation: "Static message",
      },
      withOrigin: {
        translation: "Message with origin",
        origin: [["src/App.js", 4]],
      },
      withMultipleOrigins: {
        translation: "Message with multiple origin",
        origin: [
          ["src/App.js", 4],
          ["src/Component.js", 2],
        ],
      },
    }

    const actual = format.serialize(catalog, defaultSerializeCtx)
    const pofileOriginPrefix = "#:"
    expect(actual).toEqual(expect.not.stringContaining(pofileOriginPrefix))
  })

  it("should not include lineNumbers if lineNumbers option is false", () => {
    const format = createFormatter({ origins: true, lineNumbers: false })

    const catalog: CatalogType = {
      static: {
        translation: "Static message",
      },
      withOrigin: {
        translation: "Message with origin",
        origin: [["src/App.js", 4]],
      },
      withMultipleOrigins: {
        translation: "Message with multiple origin",
        origin: [
          ["src/App.js", 4],
          ["src/Component.js", 2],
        ],
      },
    }
    const actual = format.serialize(catalog, defaultSerializeCtx)

    expect(actual).toMatchInlineSnapshot(`
      msgid ""
      msgstr ""
      "POT-Creation-Date: 2018-08-27 10:00+0000\\n"
      "MIME-Version: 1.0\\n"
      "Content-Type: text/plain; charset=utf-8\\n"
      "Content-Transfer-Encoding: 8bit\\n"
      "X-Generator: @lingui/cli\\n"
      "Language: en\\n"

      #. js-lingui-explicit-id
      msgid "static"
      msgstr "Static message"

      #. js-lingui-explicit-id
      #: src/App.js
      msgid "withOrigin"
      msgstr "Message with origin"

      #. js-lingui-explicit-id
      #: src/App.js
      #: src/Component.js
      msgid "withMultipleOrigins"
      msgstr "Message with multiple origin"

    `)
  })

  it("should not include lineNumbers if lineNumbers option is false and already excluded", () => {
    const format = createFormatter({ origins: true, lineNumbers: false })

    const catalog: CatalogType = {
      static: {
        translation: "Static message",
      },
      withOrigin: {
        translation: "Message with origin",
        origin: [["src/App.js"]],
      },
      withMultipleOrigins: {
        translation: "Message with multiple origin",
        origin: [["src/App.js"], ["src/Component.js"]],
      },
    }

    const actual = format.serialize(catalog, defaultSerializeCtx)

    expect(actual).toMatchInlineSnapshot(`
      msgid ""
      msgstr ""
      "POT-Creation-Date: 2018-08-27 10:00+0000\\n"
      "MIME-Version: 1.0\\n"
      "Content-Type: text/plain; charset=utf-8\\n"
      "Content-Transfer-Encoding: 8bit\\n"
      "X-Generator: @lingui/cli\\n"
      "Language: en\\n"

      #. js-lingui-explicit-id
      msgid "static"
      msgstr "Static message"

      #. js-lingui-explicit-id
      #: src/App.js
      msgid "withOrigin"
      msgstr "Message with origin"

      #. js-lingui-explicit-id
      #: src/App.js
      #: src/Component.js
      msgid "withMultipleOrigins"
      msgstr "Message with multiple origin"

    `)
  })
})
