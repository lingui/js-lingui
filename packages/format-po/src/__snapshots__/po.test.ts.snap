// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pofile format should correct badly used comments 1`] = `
{
  withDescriptionAndComments: {
    comments: [
      Single description only,
      Second description?,
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [
        Translator comment,
      ],
    },
    obsolete: false,
    origin: [],
    translation: Second description joins translator comments,
  },
  withMultipleDescriptions: {
    comments: [
      First description,
      Second comment,
      Third comment,
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    obsolete: false,
    origin: [],
    translation: Extra comments are separated from the first description line,
  },
}
`;

exports[`pofile format should not add lingui id more than one time 1`] = `
msgid ""
msgstr ""
"POT-Creation-Date: 2018-08-27 10:00+0000\\n"
"MIME-Version: 1.0\\n"
"Content-Type: text/plain; charset=utf-8\\n"
"Content-Transfer-Encoding: 8bit\\n"
"X-Generator: @lingui/cli\\n"
"Language: en\\n"

#. js-lingui-id: Dgzql1
msgctxt "my context"
msgid "with generated id"
msgstr ""

`;

exports[`pofile format should print lingui id if printLinguiId = true 1`] = `
msgid ""
msgstr ""
"POT-Creation-Date: 2018-08-27 10:00+0000\\n"
"MIME-Version: 1.0\\n"
"Content-Type: text/plain; charset=utf-8\\n"
"Content-Transfer-Encoding: 8bit\\n"
"X-Generator: @lingui/cli\\n"
"Language: en\\n"

#. js-lingui-id: Dgzql1
msgctxt "my context"
msgid "with generated id"
msgstr ""

`;

exports[`pofile format should read catalog in pofile format 1`] = `
{
  obsolete: {
    comments: [
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    obsolete: true,
    origin: [],
    translation: Is marked as obsolete,
  },
  static: {
    comments: [
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    obsolete: false,
    origin: [],
    translation: Static message,
  },
  veryLongString: {
    comments: [
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    obsolete: false,
    origin: [],
    translation: One morning, when Gregor Samsa woke from troubled dreams, he found himself transformed in his bed into a horrible vermin. He lay on his armour-like back, and if he lifted his head a little he could see his brown belly, slightly domed and divided by arches into stiff sections. The bedding was hardly able to cover it and seemed ready to slide off any moment. His many legs, pitifully thin compared with the size of the rest of him, waved about helplessly as he looked. "What's happened to me?" he thought. It wasn't a dream. His room, a proper human,
  },
  withComments: {
    comments: [
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [
        Translator comment,
        This one might come from developer,
      ],
    },
    obsolete: false,
    origin: [],
    translation: Support translator comments separately,
  },
  withDescription: {
    comments: [
      Description is comment from developers to translators,
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    obsolete: false,
    origin: [],
    translation: Message with description,
  },
  withMultipleOrigins: {
    comments: [
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    obsolete: false,
    origin: [
      [
        src/App.js,
        4,
      ],
      [
        src/Component.js,
        2,
      ],
    ],
    translation: Message with multiple origin,
  },
  withOrigin: {
    comments: [
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    obsolete: false,
    origin: [
      [
        src/App.js,
        4,
      ],
    ],
    translation: Message with origin,
  },
  xC8QeX: {
    comments: [
      js-lingui-id: pXy+hm,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    message: Message with default hash id,
    obsolete: false,
    origin: [],
    translation: Translation for: Message with default-hash id,
  },
  xLTujh: {
    comments: [],
    context: null,
    extra: {
      flags: [
        fuzzy,
        otherFlag,
        explicit-id,
      ],
      translatorComments: [],
    },
    message: withFlags,
    obsolete: false,
    origin: [],
    translation: Keeps any flags that are defined,
  },
}
`;

exports[`pofile format should write catalog in pofile format 1`] = `
msgid ""
msgstr ""
"POT-Creation-Date: 2018-08-27 10:00+0000\\n"
"MIME-Version: 1.0\\n"
"Content-Type: text/plain; charset=utf-8\\n"
"Content-Transfer-Encoding: 8bit\\n"
"X-Generator: @lingui/cli\\n"
"Language: en\\n"

#. js-lingui-explicit-id
msgid "static"
msgstr "Static message"

#. js-lingui-explicit-id
#: src/App.js:4
msgid "withOrigin"
msgstr "Message with origin"

#. js-lingui-explicit-id
msgctxt "my context"
msgid "withContext"
msgstr "Message with context"

msgctxt "my context"
msgid "with generated id"
msgstr ""

#. js-lingui-explicit-id
#: src/App.js:4
#: src/Component.js:2
msgid "withMultipleOrigins"
msgstr "Message with multiple origin"

#. Description is comment from developers to translators
#. js-lingui-explicit-id
msgid "withDescription"
msgstr "Message with description"

# Translator comment
# This one might come from developer
#. js-lingui-explicit-id
msgid "withComments"
msgstr "Support translator comments separately"

#. js-lingui-explicit-id
#~ msgid "obsolete"
#~ msgstr "Obsolete message"

#. js-lingui-explicit-id
#, fuzzy,otherFlag
msgid "withFlags"
msgstr "Keeps any flags that are defined"

#. js-lingui-explicit-id
msgid "veryLongString"
msgstr "One morning, when Gregor Samsa woke from troubled dreams, he found himself transformed in his bed into a horrible vermin. He lay on his armour-like back, and if he lifted his head a little he could see his brown belly, slightly domed and divided by arches into stiff sections. The bedding was hardly able to cover it and seemed ready to slide off any moment. His many legs, pitifully thin compared with the size of the rest of him, waved about helplessly as he looked. \\"What's happened to me?\\" he thought. It wasn't a dream. His room, a proper human"

`;
